web_monitoring/db.py

```diff
@@ -1,47 +1,658 @@
+        res = requests.get(url, auth=self._auth)
+        _process_errors(res)
+        result = res.json()
+        data = result['data']
+        data['capture_time'] = parse_timestamp(data['capture_time'])
+        data['updated_at'] = parse_timestamp(data['updated_at'])
+        data['created_at'] = parse_timestamp(data['created_at'])
+        return result
+
+    def add_version(self, *, page_id, capture_time, uri, hash,
+                    source_type, title, uuid=None, source_metadata=None):
+        """
+        Submit one new Version.
+
+        See :func:`add_versions` for a more efficient bulk importer.
+
+        Parameters
+        ----------
+        page_id : string
+            Page to which the Version is associated
+        uri : string
+            URI of content (such as an S3 bucket or InternetArchive URL)
+        hash : string
+            SHA256 hash of Version content
+        source_type : string
+            such as 'versionista' or 'internetarchive'
+        title : string
+            content of ``<title>`` tag
+        uuid : string, optional
+            A new, unique Version ID (UUID4). If not specified, the server
+            will generate one.
+        source_metadata : dict, optional
+            free-form metadata blob provided by source
+
+        Returns
+        -------
+        response : dict
+        """
+        # Do some type casting here as gentle error-checking.
+        version = _build_version(
+            page_id=page_id,
+            uuid=uuid,
+            capture_time=capture_time,
+            uri=uri,
+            hash=hash,
+            source_type=source_type,
+            title=title,
+            source_metadata=source_metadata)
+        url = f'{self._api_url}/pages/{page_id}/versions'
+        res = requests.post(url, auth=self._auth,
+                            headers={'Content-Type': 'application/json'},
+                            data=json.dumps(version))
+        _process_errors(res)
+        return res.json()
+
+    def add_versions(self, versions, *, update='skip', batch_size=50000):
+        """
+        Submit versions in bulk for importing into web-monitoring-db.
+
+        Chunk the versions into batches of at most the given size.
+
+        Parameters
+        ----------
+        versions : iterable
+            Iterable of dicts from :func:`format_version`
+        update : {'skip', 'replace', 'merge'}, optional
+            Specifies how versions that are already in the database (i.e.
+            versions with the same ``capture_time`` and ``source_type``) should
+            be handled:
+
+                * ``'skip'`` (default) -- Don’t import the version or modify
+                  the existing database entry.
+                * ``'replace'`` -- Replace the existing database entry with the
+                  imported one
+                * ``'merge'`` -- Similar to `replace`, but merges the values in
+                  ``source_metadata``
```

Comment:
FWIW, this will merge other fields, too, in the sense that, if they aren’t specified, they won’t be updated (unlike replace, which clears any unspecified fields in the target record). So you could use this behavior to change just the metadata or just the source type without amending the URI or hash, etc. Dunno if that matters here, though, since I think `_build_importable_version` will never create an object with unspecified fields.

Reply:
Let's leave this as is but revise the description in swagger.yml (where I cribbed this from).